### PR TITLE
ci: tweaks to CircleCI config to improve speed/reliability

### PR DESCRIPTION
**Pull request by DSchau**
Note: this is a WIP. I'm thinking CircleCI might work better for this,
basically we can report back status quickly on unit/integration tests,
and then conditionally (i.e. push button) trigger off the e2e tests
before we merge a PR. So we get the best of both worlds, relatively
quick running builds and then the assurance that we're not breaking
things before merge.

### CODE REVIEW

**Comment by m-allanson** [comment]
One thing I wondered about before, but I'm not sure how feasible it is... `yarn bootstrap` takes a significant amount of time, do you think there's a way to run that once at the beginning of the CI run, then pass the built files into the separate jobs within the CI run? My gut says it's probably more faff than it's worth, but could be a nice win if there's a clean way to do it.

**Comment by DSchau** [comment]
@m-allanson looks like we'd have to [persist to s3](https://docs.travis-ci.com/user/build-stages#data-persistence-between-stages-and-jobs), and then pull it down as part of the next job. _Could_ be worth doing, I'll take a look :)

**Comment by DSchau** [comment]
This is _close_. I can keep going with optimizing CircleCI workflows--which I think is a valuable approach--or I can switch gears a bit and get back to optimizing Travis. Thoughts appreciated!

Here's where I ended up with this:

screen shot 2018-09-18 at 5 34 12 pm

- General idea is that bootstrap/integration/lint run in parallel
- bootstrap "calls" test with its result(s) saved
- run-e2e is a manual step that must be run to trigger the e2e tests

The ultimate goal is to get more actionable feedback to our contributors more quickly. I think this achieves that, but definitely still more to do.

**Comment by m-allanson** [comment]
This looks good! I guess it'd be interesting to have near-equivalent setups running on Travis and Circle CI, then see how they're doing after a week or two?

**Comment by DSchau** [comment]
@m-allanson agreed! Don't want to call it one way or the other yet!

I was also thinking it might be valuable to have a bit more intelligent test running. So we _only_ run e2e tests and unit tests if packages/ changed. Something like that could be valuable and get feedback back more quickly!

(also you gave me this idea 👍, so thank you!)

File: scripts/assert-changed-files.sh [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
+GREP_PATTERN=$1
+
+CHANGED_FILES="$(git diff master... --name-only -r | grep -E "$GREP_PATTERN")"
```
Comment (DSchau) [inline scripts/assert-changed-files.sh]: This was _kind of_ exhibiting weird behavior on CircleCI. This looks accurate to more Unix-y 👀 than my own right?

Only thing I wasn't sure of is whether I'd need to fetch origin for this to work.

File: scripts/integration-test.sh [hunk 2]
```diff
@@ -1,18 +1,14 @@
 #!/bin/bash
-INTEGRATION_TEST=$1
-SRC_PATH=$2
-GATSBY_PATH="${TRAVIS_BUILD_DIR:-../../}" # set to third arg if defined, otherwise use ../../
+SRC_PATH=$1
+CUSTOM_COMMAND="${2:-test}"
+GATSBY_PATH="/home/circleci/project"
```
Comment (DSchau) [inline scripts/integration-test.sh]: `CIRCLE_WORKING_DIRECTORY` annoyingly includes a `~`, e.g. `~/project`

In the container, Node resolves this (for some reason) to

```js
const path = require('path');

path.resolve(`~/project`); // /home/circleci/project/integration-tests/path-prefix/~/project
```

which is a non-existant path.

Comment (DSchau) [reply scripts/integration-test.sh]: Also I tried relative directory here, but that seemed to fail too. `../../` should be the ticket but didn't quite work consistently. It's possible changed to `gatsby-dev-cli` will be necessary to fully resolve this behavior.

File: scripts/integration-test.sh [hunk 3]
```diff
@@ -1,18 +1,14 @@
...
 
-if [[ "$INTEGRATION_TEST" = true ]]; then
-  npm install -g gatsby-dev-cli
+sudo npm install -g gatsby-dev-cli &&
```
Comment (DSchau) [inline scripts/integration-test.sh]: Thanks @davidbailey00 :) Stole this 🙃

File: scripts/integration-test.sh [hunk 4]
```diff
@@ -1,18 +1,14 @@
...
-  echo "=== setting up link to current changes with gatsby-dev in $(pwd)"
-  gatsby-dev --set-path-to-repo $GATSBY_PATH
-  gatsby-dev --scan-once --quiet --copy-all # copies _all_ files in gatsby/packages
-fi
+# setting up child integration test link to gatsby packages
+cd $SRC_PATH &&
+yarn &&
+gatsby-dev --set-path-to-repo $GATSBY_PATH &&
+gatsby-dev --scan-once --copy-all && # copies _all_ files in gatsby/packages
+yarn $CUSTOM_COMMAND &&
```
Comment (DSchau) [inline scripts/integration-test.sh]: Thinking ahead. Eventually we'll want to run `yarn test:build`, `yarn test:develop`, etc.

**Pull request merged**